pkg/apis/kudo/v1beta1/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+|------------------|-----------------|---------|--------------|------------|-------|---------------------------------------------------------------------------------------------------------------------|
+| x                | x               |         |              |            | [No]  | Forbid parameter updates when a plan is scheduled unless the same plan is triggered (instance status will be reset) |
+| x                |                 | x       |              |            | No    | Forbid upgrades is another plan is running                                                                          |
+| x                |                 |         | x            |            | No    | Forbid plan overrides (for now)                                                                                     |
+| x                |                 |         |              | x          | No    | Forbid plan cancellations (for now)                                                                                 |
+| ---              |                 |         |              |            |       | ---                                                                                                                 |
+|                  | x               |         | x            |            | No    | Forbid simultaneous parameter update and directly triggered plan                                                    |
+|                  |                 | x       | x            |            | No    | Forbid simultaneous upgrades and directly triggered plans                                                           |
+|                  | x               | x       |              |            | No*   | Forbid simultaneous upgrades and parameter updates                                                                  |
+ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ *Note: simultaneous upgrade and parameter update is NOT allowed. However, there is a exception where new OV removes an existing
+ parameter. Removing this parameter in the instance update would not count as parameter update (since there is no plan to trigger).
+
+ For the complete set of rules, see the corresponding test.
+*/
+
+// validateUpdate takes in the old and new (updated) instance and returns a new plan that might
+// be triggered based on the update and an error if the update is not valid. Return plan might be
+// - <nil> when there is no change to an existing scheduled plan
+// - '' empty string when an existing plan should be canceled (not implemented yet)
+// - 'newPlan' some new plan that should be triggered
+func validateUpdate(old, new *Instance, ov *OperatorVersion) (*string, error) {
```

Comment:
if we're not modifying these should we take pointers or rather values?

Reply:
`validateUpdate` doesn't modify but both parameters are already pointers so I'll have to dereference (and pass copies). I know that the general language guideline is to pass values unless they're too big or needed to be changed. but I always saw it as a "soft guideline" rather than a rule. wdyt?